### PR TITLE
add no-self-contained option for dotnet publish

### LINKED ISSUE #9949: dotnet publish should expose --framework-dependent
## Steps to reproduce

Publish using a RID without specifying `--self-contained`:

```
dotnet publish --runtime win-x86
```

## Expected  behavior

You get a framework-dependent application with an executable to launch.

## Actual behavior

You get a self-contained application.

## Desired change

As soon as `--runtime` is specified, the default for `publish` assumes `--self-contained`. The developer can in fact get a framework dependent application by specifying `--self-contained:false`:

```
dotnet publish --runtime win-x86 --self-contained:false
```

However passing an explicit value for flag-based options is not intuitive, because it doesn't follow regular conventions for GNU-style options.

We don't believe we can change of the default ` --self-contained` when publishing with a RID due to backwards compatibility concerns, so we should at least expose a flag that allows to produce runtime-specific, but framework-dependent outputs without crazy syntax:

```
dotnet publish --runtime win-x86 --framework-dependent
```

<details>
 <summary>Environment data</summary>
.NET Core SDK (reflecting any global.json):
 Version:   2.1.401
 Commit:    91b1c13032

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.401\

Host (useful for support):
  Version: 2.1.3-servicing-26724-03
  Commit:  124038c13e

.NET Core SDKs installed:
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3-servicing-26724-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>

### PR DESCRIPTION
**Pull request by dsychin**
This is an attempt to fix #9949

I am not sure whether the way I have done it is the best way.

### CODE REVIEW

File: src/dotnet/commands/dotnet-publish/PublishCommandParser.cs [hunk 1]
```diff
@@ -49,6 +49,23 @@ internal static class PublishCommandParser
                             string value = o.Arguments.Any() ? o.Arguments.Single() : "true";
                             return $"-property:SelfContained={value}";
                         })),
+                Create.Option(
+                    "--no-self-contained",
```
Comment (peterhuene) [inline src/dotnet/commands/dotnet-publish/PublishCommandParser.cs]: I don't think we should copy the bad design of having a boolean option take an argument, like how `--self-contained` unfortunately operates (which is very un-POSIX like).

I think `--no-self-contained` should not take an option and always imply `--self-contained false` (i.e. forwarded as `-property:SelfContained=false`).

See the `--no-build` option as an example of this.

File: src/dotnet/commands/dotnet-publish/LocalizableStrings.resx [hunk 2]
```diff
@@ -155,4 +155,7 @@ The default is to publish a framework-dependent application.</value>
   <data name="CmdNoLogo" xml:space="preserve">
     <value>Do not display the startup banner or the copyright message.</value>
   </data>
+  <data name="NoSelfContainedOptionDescription" xml:space="preserve">
+    <value>Publish your application without the .NET Core runtime.</value>
```
Comment (peterhuene) [inline src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]:
```suggestion
    <value>Publish your application without the .NET Core runtime. A supported .NET Core runtime must be installed on the target system to run your application.</value>
```

@KathleenDollard for string review.

File: test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs [hunk 3]
```diff
@@ -116,6 +116,8 @@ public void ItPublishesSelfContainedWithRid(string args)
 
         [Theory]
         [InlineData("--self-contained=false")]
+        [InlineData("--no-self-contained")]
+        [InlineData("--no-self-contained=true")]
```
Comment (peterhuene) [inline test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs]: We can remove this once `--no-self-contained` takes no arguments.

File: test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs [hunk 4]
```diff
@@ -145,6 +147,8 @@ public void ItPublishesFrameworkDependentWithRid(string args)
         [Theory]
         [InlineData("--self-contained=false")]
         [InlineData(null)]
+        [InlineData("--no-self-contained")]
+        [InlineData("--no-self-contained=true")]
```
Comment (peterhuene) [inline test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs]: We can remove this once `--no-self-contained` takes no arguments.

**Comment by dsychin** [comment]
> In addition to the comments, the `--self-contained` and `--no-self-contained` should be conflicting options. The best place to make this check would be in the publish command after it parsers its command line. It should throw `GracefulException` with the error message.
> 
> If you need any guidance in making the above changes, please reach out to me via the PR or at my Microsoft email address in my GitHub profile. Thank you!

Could you give me some hint on how I would check if `--self-contained` is being used?

File: src/dotnet/commands/dotnet-publish/PublishCommandParser.cs (hunk 1 above)
Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-publish/PublishCommandParser.cs]: Agree --no-self-contained, no argument. 

```
--no-self-contained false 
```

makes my head hurt

File: src/dotnet/commands/dotnet-publish/LocalizableStrings.resx (hunk 2 above)
Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: I would like framework dependent here. How about

Publish your application as a framework dependent application without the .NET Core runtime. A supported .NET Core runtime must be installed to run your application.</value>

Comment (peterhuene) [reply src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: I like yours better, thanks!

**Comment by peterhuene** [comment]
Hi @dsychin.  Sorry for the late reply.

This would probably be the place where the conflicting argument check would go: https://github.com/dotnet/cli/blob/master/src/dotnet/commands/dotnet-publish/Program.cs#L40.

The check would be something along the lines of:

```c#
if (appliedPublishOption.HasOption("--self-contained") &&
    appliedPublishOption.HasOption("--no-self-contained"))
{
    throw new GracefulException(LocalizableStrings.SelfContainAndNoSelfContainedConflict);
}
```

We should probably have a test that covers the conflict as well.

Thanks!

**Comment by dsychin** [comment]
I am having a problem where my tests are failing.
I have changed the publish command parser to take no arguments for `--no-self-contained`.
```csharp
Create.Option(
    "--no-self-contained",
    LocalizableStrings.NoSelfContainedOptionDescription,
    Accept.NoArguments().ForwardAs("-property:SelfContained=false"))
```

Building the solution and testing it manually works, but the test is failing with 'Unknown switch' in msbuild.

```
Executing (Captured Output) - D:\Projects\cli\artifacts\tmp\Debug\test\dotnet-publish.Tests\PublishApp_win7-x64_--self-contained=false\MSBuildTestApp\bin\Debug\netcoreapp3.0\win7-x64\publish\MSBuildTestApp.exe  -
TestAsset Restore 'MSBuildTestApp'
Executing - D:\Projects\cli\artifacts\tmp\Debug\dotnet\dotnet.exe publish    -r win7-x64  --no-self-contained -  in pwd D:\Projects\cli\artifacts\tmp\Debug\test\dotnet-publish.Tests\PublishApp_win7-x64_--no-self-contained\MSBuildTestApp
[xUnit.net 00:01:50.74]     Microsoft.DotNet.Cli.Publish.Tests.GivenDotnetPublishPublishesProjects.ItPublishesFrameworkDependentWithRid(args: "--no-self-contained") [FAIL]
Failed   Microsoft.DotNet.Cli.Publish.Tests.GivenDotnetPublishPublishesProjects.ItPublishesFrameworkDependentWithRid(args: "--no-self-contained")
Error Message:
 Expected command to pass but it did not.
File Name: D:\Projects\cli\artifacts\tmp\Debug\dotnet\dotnet.exe
Arguments: publish    -r win7-x64  --no-self-contained
WorkingDir:: D:\Projects\cli\artifacts\tmp\Debug\test\dotnet-publish.Tests\PublishApp_win7-x64_--no-self-contained\MSBuildTestApp
Exit Code: 1
StdOut:
Microsoft (R) Build Engine version 16.2.0-preview-19278-01+d635043bd for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1001: Unknown switch.
Switch: --no-self-contained

For switch syntax, type "MSBuild -help"
```

File: src/dotnet/commands/dotnet-publish/LocalizableStrings.resx [hunk 5]
```diff
@@ -155,4 +155,10 @@ The default is to publish a framework-dependent application.</value>
   <data name="CmdNoLogo" xml:space="preserve">
     <value>Do not display the startup banner or the copyright message.</value>
   </data>
+  <data name="NoSelfContainedOptionDescription" xml:space="preserve">
+    <value>Publish your application as a framework dependent application without the .NET Core runtime. A supported .NET Core runtime must be installed to run your application.</value>
+  </data>
+  <data name="SelfContainAndNoSelfContainedConflict" xml:space="preserve">
+    <value>Self contained and no self contained arguments are conflicting.</value>
```
Comment (peterhuene) [inline src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: Let's use the option names in the message.  How about:

```suggestion
    <value>The '--self-contained' and '--no-self-contained' options conflict with each other.  Specify only one of the options.
```

cc @KathleenDollard for additional input.

Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: This updated message looks good

**Comment by KathleenDollard** [comment]
Do you need any more input from me on this?

**Comment by peterhuene** [comment]
@KathleenDollard no, I think that's it. Thanks for the input!

File: test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs [hunk 6]
```diff
@@ -164,6 +166,29 @@ public void ItPublishesFrameworkDependentWithoutRid(string args)
...
+                .WithRestoreFiles();
+
+            var testProjectDirectory = testInstance.Root;
+
+            new PublishCommand()
+                .WithRuntime(rid)
+                .WithWorkingDirectory(testProjectDirectory)
+                .Execute(args)
+                .Should().Fail()
+                    .And.HaveStdErr();
```
Comment (peterhuene) [inline test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs]: @dsychin whoops, sorry I missed this from my earlier review.  Could we change `HaveStdErr` to: `HaveStdErrContaining(LocalizableStrings.SelfContainAndNoSelfContainedConflict)`?

That way the test verifies that the correct error message is being output.

**Comment by peterhuene** [comment]
@dsychin thank you very much for your contribution!

**Pull request merged**